src/RecoveryServices/RecoveryServices.Backup/Cmdlets/Item/UndoAzureRmRecoveryServicesBackupItemDeletion.cs

```diff
@@ -0,0 +1,78 @@
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ----------------------------------------------------------------------------------
+
+using Microsoft.Azure.Commands.RecoveryServices.Backup.Cmdlets.Models;
+using Microsoft.Azure.Commands.RecoveryServices.Backup.Cmdlets.ProviderModel;
+using Microsoft.Azure.Commands.RecoveryServices.Backup.Properties;
+using Microsoft.Azure.Management.Internal.Resources.Utilities.Models;
+using System.Collections.Generic;
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.RecoveryServices.Backup.Cmdlets
+{
+    /// <summary>
+    /// Rehydrates a backup item from softdeleted state
+    /// </summary>
+    [Cmdlet("Undo", ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "RecoveryServicesBackupItemDeletion", SupportsShouldProcess = true), OutputType(typeof(JobBase))]
+    public class UndoAzureRmRecoveryServicesBackupItemDeletion : RSBackupVaultCmdletBase
+    {
+        /// <summary>
+        /// The protected item which needs to be rehydrated
+        /// </summary>
+        [Parameter(Position = 1, Mandatory = true, HelpMessage = ParamHelpMsgs.Item.ProtectedItem,
+            ValueFromPipeline = true)]
+        [ValidateNotNullOrEmpty]
+        public ItemBase Item { get; set; }
+
+        /// <summary>
+        /// Prevents the confirmation dialog when specified.
+        /// </summary>
+        [Parameter(Mandatory = false, HelpMessage = ParamHelpMsgs.Item.ForceOption)]
+        public SwitchParameter Force { get; set; }
+
+        public override void ExecuteCmdlet()
+        {
+
+            ExecutionBlock(() =>
+            {
+                base.ExecuteCmdlet();
+
+                ResourceIdentifier resourceIdentifier = new ResourceIdentifier(VaultId);
+                string vaultName = resourceIdentifier.ResourceName;
+                string resourceGroupName = resourceIdentifier.ResourceGroupName;
+
+                PsBackupProviderManager providerManager =
+                    new PsBackupProviderManager(new Dictionary<System.Enum, object>()
+                    {
+                        { VaultParams.VaultName, vaultName },
+                        { VaultParams.ResourceGroupName, resourceGroupName },
+                        { ItemParams.Item, Item },
+                    }, ServiceClientAdapter);
+
+                IPsBackupProvider psBackupProvider =
+                    providerManager.GetProviderInstance(Item.WorkloadType,
+                    Item.BackupManagementType);
+
+                var itemResponse = psBackupProvider.UndeleteProtection();
```

Comment:
So, what is tp prevent this from calling one of the unimplemented methods for UndeleteProtection, given that you only have one class that implements this.

At least, you need to catch the NotImplementedException that would occur here for most of your providers  and return a message letting the user know that this is not supported..  At best, you should have another indicator of whether the method is implemented (for example, having a separate interface for undelete, or making the method implenent a TryUndeleteMethod that returns false if Undelete is not possible, rather than throwing an uncaught exception)..

Reply:
This is the coding style we are following as of now. There are other features as well which are only enabled for some specific provider. For example, **RegisterContainer()** method is only implemented for Workload Provider. If called for other providers, the same behavior is expected. i.e. to throw NotImplementedException

Reply:
@markcowl The undelete feature for other workloads will arrive as and when they are supported by our service which is planned to come in the near future.

Reply:
This is not an acceptable customer experience.  You *must* make the change descried above in order to have this PR merged.